From 06f960d74e299b3cb7d95e05577278a94f7d43f2 Mon Sep 17 00:00:00 2001
From: Oliver Bestmann <oliver.bestmann@googlemail.com>
Date: Tue, 16 Dec 2025 11:59:22 +0100
Subject: [PATCH] drm: apple: set timestamps for 120hz

After looking into the timestamps sent by macos in dcp_swap_submit, it seems to
me that ts3 is strictly increasing and always set. ts1 and ts2 are set
a few milliseconds before ts3 (about 4ms) and ts1 a few milliseconds after
ts3 (about 16ms).

All three timestamps are in mach absolute time (i.e. 41.67ns per timestep).

MacOS also sets flags1 and flags2. I have no idea what exactly those values
indicate, but I did do not need to set any of them to get 120hz.
---
 drivers/gpu/drm/apple/iomfb_template.c | 13 +++++++++++++
 drivers/gpu/drm/apple/iomfb_template.h |  9 ++++++++-
 drivers/gpu/drm/apple/parser.c         | 13 -------------
 3 files changed, 21 insertions(+), 14 deletions(-)

diff --git a/drivers/gpu/drm/apple/iomfb_template.c b/drivers/gpu/drm/apple/iomfb_template.c
index 0a1b9495128562..a382c53b711c90 100644
--- a/drivers/gpu/drm/apple/iomfb_template.c
+++ b/drivers/gpu/drm/apple/iomfb_template.c
@@ -1264,6 +1264,10 @@ int DCP_FW_NAME(iomfb_modeset)(struct apple_dcp *dcp,
 	return 0;
 }
 
+static u64 convert_ns_to_mach_time(u64 time) {
+	return time * 100 / 4167;
+}
+
 void DCP_FW_NAME(iomfb_flush)(struct apple_dcp *dcp, struct drm_crtc *crtc, struct drm_atomic_state *state)
 {
 	struct drm_plane *plane;
@@ -1408,6 +1412,15 @@ void DCP_FW_NAME(iomfb_flush)(struct apple_dcp *dcp, struct drm_crtc *crtc, stru
 		req->clear = 1;
 	}
 
+	if (has_surface) {
+		u64 now = ktime_get_ns();
+
+		/* set timestamps to reach 120 fps */
+		req->swap.ts1 = convert_ns_to_mach_time(now + 16660000);
+		req->swap.ts2 = convert_ns_to_mach_time(now - 40000000);
+		req->swap.ts3 = convert_ns_to_mach_time(now);
+	}
+
 	/* These fields should be set together */
 	req->swap.swap_completed = req->swap.swap_enabled;
 
diff --git a/drivers/gpu/drm/apple/iomfb_template.h b/drivers/gpu/drm/apple/iomfb_template.h
index f446a4d8f38b90..878ffdf9e644cd 100644
--- a/drivers/gpu/drm/apple/iomfb_template.h
+++ b/drivers/gpu/drm/apple/iomfb_template.h
@@ -18,7 +18,14 @@
 struct DCP_FW_NAME(dcp_swap) {
 	u64 ts1;
 	u64 ts2;
-	u64 unk_10[6];
+
+	u64 unk_10;
+	u64 unk_18;
+	u64 ts64_unk;
+	u64 unk_28;
+	u64 ts3;
+	u64 unk_38;
+
 	u64 flags1;
 	u64 flags2;
 
diff --git a/drivers/gpu/drm/apple/parser.c b/drivers/gpu/drm/apple/parser.c
index 700a31883eac8e..e3d2bf824a59b2 100644
--- a/drivers/gpu/drm/apple/parser.c
+++ b/drivers/gpu/drm/apple/parser.c
@@ -499,19 +499,6 @@ static int parse_mode(struct dcp_parse_ctx *handle,
 	if (is_virtual)
 		return -EINVAL;
 
-	/*
-	 * HACK:
-	 * Ignore the 120 Hz mode on j314/j316 (identified by resolution).
-	 * DCP limits normal swaps to 60 Hz anyway and the 120 Hz mode might
-	 * cause choppiness with X11.
-	 * Just downscoring it and thus making the 60 Hz mode the preferred mode
-	 * seems not enough for some user space.
-	 */
-	if (vert.precise_sync_rate >> 16 == 120 &&
-	    ((horiz.active == 3024 && vert.active == 1964) ||
-	     (horiz.active == 3456 && vert.active == 2234)))
-		return -EINVAL;
-
 	vert.active -= notch_height;
 	vert.sync_width += notch_height;
 
